cpp/src/arrow/array/util.cc

```diff
@@ -77,7 +77,9 @@ class ArrayDataWrapper {
 
 class ArrayDataEndianSwapper {
  public:
-  explicit ArrayDataEndianSwapper(const std::shared_ptr<ArrayData>& data) : data_(data) {
+  explicit ArrayDataEndianSwapper(const std::shared_ptr<ArrayData>& data,
+                                  MemoryPool* pool)
+      : data_(data), pool_(pool) {
```

Comment:
Did you grep for all calls to `ArrayDataEndianSwapper` and explicitly pass a pool or the `default_memory_pool()` explicitly?

Context: https://github.com/apache/arrow/issues/36360

Reply:
Quick answer: yes, `ArrayDataEndianSwapper` is merly used. I've capture all callings.

And `pool_` haven't been default initialized. So compiler will prevent from that calling

Reply:
> So compiler will prevent from that calling

But you have `= default_memory_pool()` as a default value argument, no?

Reply:
1. class `ArrayDataEndianSwapper` has prevent from it
2. function `SwapEndianArrayData` doesn't. It's only used in `ipc`, and I've changed them. You can check it again

Reply:
No need to check. I just wanted to confirm. :-) PR is approved.

Reply:
Fall sleep now, thanks a lot!